compiler/src/main/scala/org/scalajs/core/compiler/PrepJSInterop.scala

```diff
@@ -865,14 +888,53 @@ abstract class PrepJSInterop extends plugins.PluginComponent
               "must be final")
         }
 
-        // Traits must be pure interfaces
+        // Traits must be pure interfaces, except for js.undefined members
         if (sym.owner.isTrait && sym.isTerm && !sym.isConstructor) {
-          if (!sym.isDeferred) {
-            reporter.error(tree.pos,
-                "A Scala.js-defined JS trait can only contain abstract members")
-          } else if (isPrivateMaybeWithin(sym)) {
+          if (sym.isMethod && isPrivateMaybeWithin(sym)) {
             reporter.error(tree.pos,
                 "A Scala.js-defined JS trait cannot contain private members")
+          } else if (!sym.isDeferred) {
+            /* Tell the back-end not emit this thing. In fact, this only
+             * matters for mixed-in members created from this member.
+             */
+            sym.addAnnotation(JSOptionalAnnotation)
+
+            def rhsPos = if (tree.rhs != EmptyTree) tree.rhs.pos else tree.pos
```

Comment:
When the method is abstract. But of course that can't happen here because we're in `if (!sym.isDeferred)`. I must have copy-pasted that from somewhere else.